File: sol/solution/0101-0200/0105/Solution.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {
    /**
     * @param Integer[] $preorder
     * @param Integer[] $inorder
     * @return TreeNode
     */
    private $map;
    private $idx = 0;
    function buildTree($preorder, $inorder) {
        $this->map = array_flip($inorder);
        return $this->helper($preorder, 0, count($inorder) - 1);
    }
    function helper(&$preorder, $left, $right) {
        if ($left > $right) return null;
        $val = $preorder[$this->idx++];
        $node = new TreeNode($val);
        $node->left = $this->helper($preorder, $left, $this->map[$val] - 1);
        $node->right = $this->helper($preorder, $this->map[$val] + 1, $right);
        return $node;
    }
}

File: sol/solution/0101-0200/0105/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
  @index = 0
  @map = {}
  inorder.each_with_index { |val, i| @map[val] = i }
  def helper(preorder, left, right)
    return nil if left > right
    val = preorder[@index]
    @index += 1
    node = TreeNode.new(val)
    node.left = helper(preorder, left, @map[val] - 1)
    node.right = helper(preorder, @map[val] + 1, right)
    node
  end
  helper(preorder, 0, inorder.length - 1)
end
